File: src-tauri/src/apis/omdb.rs
use serde::Deserialize;
use super::client::{ApiClient, ApiError};
use super::types::{MovieSearchResult, TvSearchResult, ContentDetails};

const OMDB_BASE_URL: &str = "https://www.omdbapi.com/";

#[derive(Debug, Deserialize)]
struct OmdbSearchResponse {
    #[serde(rename = "Search")]
    search: Option<Vec<OmdbSearchItem>>,
    #[serde(rename = "Response")]
    response: String,
    #[serde(rename = "Error")]
    error: Option<String>,
    #[serde(rename = "totalResults")]
    total_results: Option<String>,
}

#[derive(Debug, Deserialize)]
struct OmdbSearchItem {
    #[serde(rename = "Title")]
    title: String,
    #[serde(rename = "Year")]
    year: String,
    #[serde(rename = "imdbID")]
    imdb_id: String,
    #[serde(rename = "Type")]
    media_type: String,
    #[serde(rename = "Poster")]
    poster: String,
}

#[derive(Debug, Deserialize)]
struct OmdbDetailResponse {
    #[serde(rename = "Title")]
    title: String,
    #[serde(rename = "Year")]
    year: String,
    #[serde(rename = "Rated")]
    rated: Option<String>,
    #[serde(rename = "Released")]
    released: Option<String>,
    #[serde(rename = "Runtime")]
    runtime: Option<String>,
    #[serde(rename = "Genre")]
    genre: Option<String>,
    #[serde(rename = "Director")]
    director: Option<String>,
    #[serde(rename = "Writer")]
    writer: Option<String>,
    #[serde(rename = "Actors")]
    actors: Option<String>,
    #[serde(rename = "Plot")]
    plot: Option<String>,
    #[serde(rename = "Language")]
    language: Option<String>,
    #[serde(rename = "Country")]
    country: Option<String>,
    #[serde(rename = "Awards")]
    awards: Option<String>,
    #[serde(rename = "Poster")]
    poster: Option<String>,
    #[serde(rename = "imdbRating")]
    imdb_rating: Option<String>,
    #[serde(rename = "imdbVotes")]
    imdb_votes: Option<String>,
    #[serde(rename = "imdbID")]
    imdb_id: String,
    #[serde(rename = "Type")]
    media_type: String,
    #[serde(rename = "totalSeasons")]
    total_seasons: Option<String>,
    #[serde(rename = "Metascore")]
    metascore: Option<String>,
    #[serde(rename = "BoxOffice")]
    box_office: Option<String>,
    #[serde(rename = "Production")]
    production: Option<String>,
    #[serde(rename = "Response")]
    response: String,
    #[serde(rename = "Error")]
    error: Option<String>,
}

pub struct OmdbApi;

impl OmdbApi {
    /// Search for movies
    pub async fn search_movies(
        client: &ApiClient,
        api_key: &str,
        query: &str,
        year: Option<&str>,
    ) -> Result<Vec<MovieSearchResult>, ApiError> {
        let mut url = format!(
            "{}?apikey={}&s={}&type=movie",
            OMDB_BASE_URL,
            api_key,
            urlencoding::encode(query)
        );

        if let Some(y) = year {
            url.push_str(&format!("&y={}", y));
        }

        let response = client.get("omdb", &url).await?;
        let data: OmdbSearchResponse = response.json().await
            .map_err(|e| ApiError::InvalidResponse(e.to_string()))?;

        if data.response == "False" {
            if let Some(error) = data.error {
                if error.contains("not found") || error.contains("No results") {
                    return Ok(vec![]);
                }
                return Err(ApiError::InvalidResponse(error));
            }
            return Ok(vec![]);
        }

        let results = data.search.unwrap_or_default()
            .into_iter()
            .map(|item| MovieSearchResult {
                title: item.title,
                year: item.year,
                imdb_id: item.imdb_id,
                media_type: item.media_type,
                poster: if item.poster == "N/A" { None } else { Some(item.poster) },
            })
            .collect();

        Ok(results)
    }

    /// Search for TV shows
    pub async fn search_tv(
        client: &ApiClient,
        api_key: &str,
        query: &str,
        year: Option<&str>,
    ) -> Result<Vec<TvSearchResult>, ApiError> {
        let mut url = format!(
            "{}?apikey={}&s={}&type=series",
            OMDB_BASE_URL,
            api_key,
            urlencoding::encode(query)
        );

        if let Some(y) = year {
            url.push_str(&format!("&y={}", y));
        }

        let response = client.get("omdb", &url).await?;
        let data: OmdbSearchResponse = response.json().await
            .map_err(|e| ApiError::InvalidResponse(e.to_string()))?;

        if data.response == "False" {
            if let Some(error) = data.error {
                if error.contains("not found") || error.contains("No results") {
                    return Ok(vec![]);
                }
                return Err(ApiError::InvalidResponse(error));
            }
            return Ok(vec![]);
        }

        let results = data.search.unwrap_or_default()
            .into_iter()
            .map(|item| TvSearchResult {
                title: item.title,
                year: item.year,
                imdb_id: item.imdb_id,
                media_type: item.media_type,
                poster: if item.poster == "N/A" { None } else { Some(item.poster) },
            })
            .collect();

        Ok(results)
    }

    /// Get detailed information for a movie or TV show by IMDb ID
    pub async fn get_by_imdb_id(
        client: &ApiClient,
        api_key: &str,
        imdb_id: &str,
    ) -> Result<ContentDetails, ApiError> {
        let url = format!(
            "{}?apikey={}&i={}&plot=full",
            OMDB_BASE_URL,
            api_key,
            imdb_id
        );

        let response = client.get("omdb", &url).await?;
        let data: OmdbDetailResponse = response.json().await
            .map_err(|e| ApiError::InvalidResponse(e.to_string()))?;

        if data.response == "False" {
            if let Some(error) = data.error {
                return Err(ApiError::InvalidResponse(error));
            }
            return Err(ApiError::NotFound);
        }

        // Convert N/A values to None
        let clean_option = |opt: Option<String>| -> Option<String> {
            opt.filter(|s| s != "N/A" && !s.is_empty())
        };

        Ok(ContentDetails {
            title: data.title,
            year: data.year,
            rated: clean_option(data.rated),
            released: clean_option(data.released),
            runtime: clean_option(data.runtime),
            genre: clean_option(data.genre),
            director: clean_option(data.director),
            writer: clean_option(data.writer),
            actors: clean_option(data.actors),
            plot: clean_option(data.plot),
            language: clean_option(data.language),
            country: clean_option(data.country),
            awards: clean_option(data.awards),
            poster: clean_option(data.poster),
            imdb_rating: clean_option(data.imdb_rating),
            imdb_votes: clean_option(data.imdb_votes),
            imdb_id: data.imdb_id,
            media_type: data.media_type,
            total_seasons: clean_option(data.total_seasons),
            metascore: clean_option(data.metascore),
            box_office: clean_option(data.box_office),
            production: clean_option(data.production),
        })
    }
}

File: scripts/content/core/api/omdb.ts
/**
 * OMDB API client for movie/TV search and details
 * Uses native fetch - works in both Node.js and browser
 */

import type { MovieSearchResult, TvSearchResult, ContentDetails } from '../types.js';

const OMDB_BASE_URL = 'https://www.omdbapi.com/';

interface OmdbSearchResponse {
	Search?: OmdbSearchItem[];
	Response: string;
	Error?: string;
	totalResults?: string;
}

interface OmdbSearchItem {
	Title: string;
	Year: string;
	imdbID: string;
	Type: string;
	Poster: string;
}

interface OmdbDetailResponse {
	Title: string;
	Year: string;
	Rated?: string;
	Released?: string;
	Runtime?: string;
	Genre?: string;
	Director?: string;
	Writer?: string;
	Actors?: string;
	Plot?: string;
	Language?: string;
	Country?: string;
	Awards?: string;
	Poster?: string;
	imdbRating?: string;
	imdbVotes?: string;
	imdbID: string;
	Type: string;
	totalSeasons?: string;
	Metascore?: string;
	BoxOffice?: string;
	Production?: string;
	Response: string;
	Error?: string;
}

/**
 * Clean optional string value - convert "N/A" to undefined
 */
function cleanOption(value: string | undefined): string | undefined {
	if (!value || value === 'N/A' || value.trim() === '') {
		return undefined;
	}
	return value;
}

/**
 * Search for movies by title
 */
export async function searchMovies(
	apiKey: string,
	query: string,
	year?: string
): Promise<MovieSearchResult[]> {
	const params = new URLSearchParams({
		apikey: apiKey,
		s: query,
		type: 'movie'
	});

	if (year) {
		params.set('y', year);
	}

	const url = `${OMDB_BASE_URL}?${params.toString()}`;
	const response = await fetch(url);

	if (!response.ok) {
		throw new Error(`OMDB API error: ${response.status} ${response.statusText}`);
	}

	const data: OmdbSearchResponse = await response.json();

	if (data.Response === 'False') {
		if (data.Error?.includes('not found') || data.Error?.includes('No results')) {
			return [];
		}
		throw new Error(`OMDB API error: ${data.Error}`);
	}

	return (data.Search || []).map((item) => ({
		title: item.Title,
		year: item.Year,
		imdbId: item.imdbID,
		mediaType: item.Type,
		poster: item.Poster === 'N/A' ? undefined : item.Poster
	}));
}

/**
 * Search for TV series by title
 */
export async function searchTv(
	apiKey: string,
	query: string,
	year?: string
): Promise<TvSearchResult[]> {
	const params = new URLSearchParams({
		apikey: apiKey,
		s: query,
		type: 'series'
	});

	if (year) {
		params.set('y', year);
	}

	const url = `${OMDB_BASE_URL}?${params.toString()}`;
	const response = await fetch(url);

	if (!response.ok) {
		throw new Error(`OMDB API error: ${response.status} ${response.statusText}`);
	}

	const data: OmdbSearchResponse = await response.json();

	if (data.Response === 'False') {
		if (data.Error?.includes('not found') || data.Error?.includes('No results')) {
			return [];
		}
		throw new Error(`OMDB API error: ${data.Error}`);
	}

	return (data.Search || []).map((item) => ({
		title: item.Title,
		year: item.Year,
		imdbId: item.imdbID,
		mediaType: item.Type,
		poster: item.Poster === 'N/A' ? undefined : item.Poster
	}));
}

/**
 * Get detailed information for a movie or TV show by IMDb ID
 */
export async function getContentDetails(
	apiKey: string,
	imdbId: string
): Promise<ContentDetails> {
	const params = new URLSearchParams({
		apikey: apiKey,
		i: imdbId,
		plot: 'full'
	});

	const url = `${OMDB_BASE_URL}?${params.toString()}`;
	const response = await fetch(url);

	if (!response.ok) {
		throw new Error(`OMDB API error: ${response.status} ${response.statusText}`);
	}

	const data: OmdbDetailResponse = await response.json();

	if (data.Response === 'False') {
		if (data.Error) {
			throw new Error(`OMDB API error: ${data.Error}`);
		}
		throw new Error('Movie not found');
	}

	return {
		title: data.Title,
		year: data.Year,
		rated: cleanOption(data.Rated),
		released: cleanOption(data.Released),
		runtime: cleanOption(data.Runtime),
		genre: cleanOption(data.Genre),
		director: cleanOption(data.Director),
		writer: cleanOption(data.Writer),
		actors: cleanOption(data.Actors),
		plot: cleanOption(data.Plot),
		language: cleanOption(data.Language),
		country: cleanOption(data.Country),
		awards: cleanOption(data.Awards),
		poster: cleanOption(data.Poster),
		imdbRating: cleanOption(data.imdbRating),
		imdbVotes: cleanOption(data.imdbVotes),
		imdbId: data.imdbID,
		mediaType: data.Type,
		totalSeasons: cleanOption(data.totalSeasons),
		metascore: cleanOption(data.Metascore),
		boxOffice: cleanOption(data.BoxOffice),
		production: cleanOption(data.Production)
	};
}

// Alias for backwards compatibility
export const getMovieDetails = getContentDetails;
